data/pools/transactionPool.go

```diff
@@ -558,6 +568,19 @@ func (pool *TransactionPool) OnNewBlock(block bookkeeping.Block, delta ledger.St
 	}
 }
 
+// isAssemblyTimedOut determines if we should keep attempting complete the block assembly by adding more transactions to the pending evaluator,
+// or whether we've ran out of time. It takes into consideration the assemblyDeadline that was set by the AssembleBlock function as well as the
+// projected time it's going to take to call the GenerateBlock function before the block assembly would be ready.
+// The function expects that the pool.assemblyMu lock would be taken before being called.
+func (pool *TransactionPool) isAssemblyTimedOut() bool {
+	if pool.assemblyDeadline.IsZero() {
+		// we have no deadline, so no reason to timeout.
+		return false
+	}
+	generateBlockDuration := generateBlockBaseDuration + time.Duration(pool.pendingBlockEvaluator.TxnCounter())*generateBlockTransactionDuration
+	return time.Now().After(pool.assemblyDeadline.Add(-generateBlockDuration))
```

Comment:
Why is this '-' generatBlockDuration?

Reply:
I think we already covered this in another PR.. the `time.Add` returns a `time.Time`; the `time.Sub` returns a `time.Duration`. If we want to subtract a duration from a `time.Time` and get a `time.Time`, the correct function is the `time.Add` and passing a negative `time.Duration`.

Reply:
My concern is something else. 
The way I understand this is that generateBlockDuration is based on generateBlockTransactionDuration, which I assume, is an upper limit.
And using the upper limit for subtracting from the allowed time, might be subtracting too much, and triggering a timeout.

Reply:
These numbers were tuned by trial and error; the block assembly time in the case of a timeout, even with these changes is always more than 250ms, suggesting that these values are well within reason.
Note that if you'll do the math, you'll see that the `generateBlockDuration` never ends up being more than 13ms on a block of a 4650 transactions ( which is our current limit ).

This change is important as we're trying to push for higher transaction count, and need to have a better accuracy when stopping to build the block.